Products release 1/1775 update shipping

nbradbury opened the pull request:
Tackles the initial UI work for #1775 by adding the ability to edit product shipping settings. Like @anitaa1990's PRs for updating products, this doesn't actually update the product draft at this point - it simply provides the UI for it.

The UI consists of a tappable area containing the shipping info:

![detail](https://user-images.githubusercontent.com/3903757/73027425-5e86bf80-3e01-11ea-8110-398d9691574b.png)

A screen for editing that information:

![data](https://user-images.githubusercontent.com/3903757/73027493-870eb980-3e01-11ea-940e-cbb7cddbc0c2.png)

And a dialog for choosing the shipping class:

![dialog](https://user-images.githubusercontent.com/3903757/73027532-9aba2000-3e01-11ea-84d0-d3e588c9e7fc.png)

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Garance91540 commented:
Hi @nbradbury , looking good!

I only have a few notes on the screenshots you shared:
- I'm seeing chevrons on the "Inventory" and "Shipping" list items, but I'm not seeing them on @kyleaparker 's original designs. Is there a specific reason why we added them here?

<img width="276" alt="Screen Shot 2020-01-24 at 4 41 23 PM" src="https://user-images.githubusercontent.com/28450084/73081803-8bea7080-3ec8-11ea-826a-7dd73c600936.png">

- Once you click a product in the product list, the bottom nav bar should disappear from all screens. 

I downloaded the apk but couldn't select the shipping list item. Just want to verify with you that once a change is made in the shipping setting screens
1. users can save it by selecting the "DONE" button on the top-right corner of the screen
2. users can discard the change by selecting the back button and then "discard" in the feedback modal?

<img width="551" alt="Screen Shot 2020-01-24 at 4 55 21 PM" src="https://user-images.githubusercontent.com/28450084/73082771-56df1d80-3eca-11ea-99c1-8063748913a1.png">

Let me know if you have any questions!

nbradbury commented:
> * I'm seeing chevrons on the "Inventory" and "Shipping" list items, but I'm not seeing them on @kyleaparker 's original designs. Is there a specific reason why we added them here?

Ah, right. We added those temporarily to indicate which fields are editable. This feature is only enabled in debug builds, and the chevrons will be removed before it's released.

> Once you click a product in the product list, the bottom nav bar should disappear from all screens.

Hmm, that's not related to this PR but note that this is how it works throughout the app. Clicking an order in the order list, for example, doesn't make the nav bar disappear. If this should be changed let me know and I'll file an issue for it.

>  users can save it by selecting the "DONE" button on the top-right corner of the screen (and) users can discard the change by selecting the back button and then "discard" in the feedback modal?

This will be added, but at this point we don't actually save any changes. I noticed in @anitaa1990's previous work on updating products that she decided to leave the actual updating to a later PR and focus on the UI.

Garance91540 commented:
Hi @nbradbury thank you for your answers, all makes sense now on my end 👍

> Once you click a product in the product list, the bottom nav bar should disappear from all screens.
I looked around at different apps and it seems like there's not a single common pattern. Let's leave it as we have it!

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductShippingFragment.kt:
```diff
@@ -0,0 +1,164 @@
+
+    private fun initializeViewModel() {
+        setupObservers(viewModel)
+        viewModel.start(navArgs.remoteProductId)
```
Using the `start()` initialization is no longer necessary because you can access navigation arguments through `savedState.navArgs()` in the ViewModel -- so you can perform all the VM initialization in the constructor. This is preferred, especially in the new features, because it avoids the potential `UninitializedPropertyAccessException`.

nbradbury replied:
Thanks, I somehow missed that! Fixed in 0c2d6d3.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductShippingFragment.kt:
```diff
@@ -0,0 +1,164 @@
+
+    private fun showShippingClassDialog() {
+        isShippingClassDlgShowing = true
+        shippingClassDialog = ProductShippingClassDialog.newInstance(this@ProductShippingFragment).also { dialog ->
```
I don't think I knew that, thanks! I just converted over showing the dialog via the nav controller, but there doesn't appear to be a way to set the dialog's target fragment. This is necessary so the shipping dialog can call the listener in the shipping fragment.

I suppose I could use a shared `ViewModel` but that seems overkill, so if it's okay with you I will just leave this as-is.

nbradbury commented:
> By the way, is this meant to be merged to `develop` even though it's not functional?

@0nko Yes, this is fine. It's hidden by a feature flag and only appears in debug builds.

0nko commented:
Thanks for the changes @nbradbury! I just noticed one more thing - when you rotate the screen, the shipping class selection is reset. If you were planning to address this in a future PR we can merge this.

anitaa1990 commented:
@nbradbury, not sure if this is going to be addressed in another PR but just thought I would mention it here. The shipping classes are a paginated list and we might need to fetch more when the user scrolls. @kyleaparker had mentioned that we would be changing the shipping class list from a list dialog to it's own separate screen, something similar to the pattern we will be using on the Product Settings screens. iOS is also implemented the same way.

(LEFT: Figma design for product settings page . RIGHT: iOS implementation)
<img width="300" src="https://user-images.githubusercontent.com/22608780/73355215-9c7e5a80-42bd-11ea-91b1-2c6803914555.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/73355225-a4d69580-42bd-11ea-9d90-bc28e9d5ca81.png">

nbradbury commented:
> The shipping classes are a paginated list and we might need to fetch more when the user scrolls.

@anitaa1990 This is already supported in this PR.

> mentioned that we would be changing the shipping class list from a list dialog to it's own separate screen

Oh, I wasn't aware of this design. I think I'd prefer to make that change in a separate PR, if that's okay with you.

nbradbury commented:
> when you rotate the screen, the shipping class selection is reset. If you were planning to address this in a future PR we can merge this.

@0nko Yeah, let's address this down the road.
